loki/transformations/sanitise/associates.py

```diff
@@ -143,17 +157,31 @@ def map_scalar(self, expr, *args, **kwargs):
         return expr.clone(scope=scope.parent)
 
     def map_array(self, expr, *args, **kwargs):
-        """ Special case for arrys: we need to preserve the dimensions """
-        new = self.map_variable_symbol(expr, *args, **kwargs)
+        """ Partially resolve dimension indices and handle shape """
+
+        # Recurse over existing array dimensions
+        expr_dims = self.rec(expr.dimensions, *args, **kwargs)
 
         # Recurse over the type's shape
         _type = expr.type
         if expr.type.shape:
             new_shape = self.rec(expr.type.shape, *args, **kwargs)
             _type = expr.type.clone(shape=new_shape)
 
+        # Stop if scope is not an associate
+        if not isinstance(expr.scope, ir.Associate):
+            return expr.clone(dimensions=expr_dims, type=_type)
+
+        new = self.map_scalar(expr, *args, **kwargs)
+
         # Recurse over array dimensions
-        new_dims = self.rec(expr.dimensions, *args, **kwargs)
+        if isinstance(new, sym.Array) and new.dimensions:
+            # Resolve unbound range symbols form existing indices
+            new_dims = self.rec(new.dimensions, *args, **kwargs)
+            new_dims = self._match_range_indices(new_dims, expr_dims)
+        else:
+            new_dims = expr_dims
+
```

Comment:
I'm not sure I'm following the logic here. The first recursion to expr_dims seems to only be used if the array's scope isn't an associate - could we not move it into the conditional just before the return statement?

Or alternatively, in which situation do we expect the return value of the recursion on `new.dimensions` to differ from the first recursion? Could we maybe use the previous recursion value here?

I suspect the second wouldn't work but the first should allow us to save some recursion maybe?

Reply:
Ah, yes, this was quite fiddly. The initial recursion is required for resolving associations in the dimensions of arrays that are not actually scope to the `Associate`; that's also why it has to happen before the short-circuiting. For example, consider 
```
associate(a => b)
  some%obj(i, :, a) = ...
end associate
```

The second recursion is then done on the potentially replaced symbol (`new`), which now might have entirely different `.dimensions`. For example consider
```
associate(a => b(:, : i))
  a = 42.0
  ! or even
  a(j, k) = 66.6
end associate
```
In the first line of this example `expr == 'a'` and `expr.dimensions == None` and `new.dimensions == (:, :, i)`, the latter of which we have not recursed into yet. In the second example line `new.dimensions` then also needs to be matched against `(j, k)` to yield `b(j, k, i)`. Does that make sense?

Reply:
Thanks, I suspected something along these lines.

My question was mostly on the grounds of avoiding redundant recursion, purely judged from the dependencies within this method. Something like this:

```python
        # Recurse over the type's shape
        _type = expr.type
        if _type.shape:
            new_shape = self.rec(_type.shape, *args, **kwargs)
            _type = expr.type.clone(shape=new_shape)

        # Stop if scope is not an associate, but still recurse on the dimensions
        # in case an associated symbol is in there
        if not isinstance(expr.scope, ir.Associate):
            expr_dims = self.rec(expr.dimensions, *args, **kwargs)
            return expr.clone(dimensions=expr_dims, type=_type)

        # For arrays that represent an associate symbol, re-use the
        # scalar implementation to resolve the symbol
        new = self.map_scalar(expr, *args, **kwargs)

        # Recurse over array dimensions
        if isinstance(new, sym.Array) and new.dimensions:
            # Resolve unbound range symbols form existing indices
            new_dims = self.rec(new.dimensions, *args, **kwargs)
            new_dims = self._match_range_indices(new_dims, expr_dims)
        else:
            new_dims = expr_dims
        return new.clone(dimensions=new_dims, type=_type)
```

This should save the need to do the recursion on the dimensions for symbols from the associate since you're recursing on the `new_dims` anyway